### ISSUE TITLE
PLC3002: false positive

### ISSUE THREAD

**Issue by spaceone** [astral-sh/ruff#19135]
### Summary

```python
import re


class language:
    choices = [ 
        ('', ''),
        ('af_ZA', 'Afrikaans/South Africa'),
        ('af_ZA.UTF-8', 'Afrikaans/South Africa(UTF-8)'),
        ('sq_AL', 'Albanian/Albania'),
    ] 


class languageCode:
    _re = re.compile(r'^[a-z][a-z]_[A-Z][A-Z]$')
    choices = (lambda m: [kv for kv in language.choices if m(kv[0])])(_re.match)  # noqa: PLC3002
```

contains `PLC3002` (last line) while it cannot be written as suggested:

```python
class languageCode:
    _re = re.compile(r'^[a-z][a-z]_[A-Z][A-Z]$')
    choices = [kv for kv in language.choices if _re.match(kv[0])]
```
→ causes `F821: Undefined name '_re'`

### Version

ruff 0.11.13

**Comment by MeGaGiGaGon** [comment]
Here's a much simpler example: https://play.ruff.rs/aa033d95-2050-4e59-9fa9-27438634a87f
```py
class A:
    x = 1
    y = [_ for _ in [1] if x]  # F821
    y = (lambda test:[_ for _ in [1] if test])(x)  # PLC3002
```

**Comment by ntBre** [comment]
Wow, another scoping quirk in classes. So `x` is in scope for the call, but not in the comprehension?

It looks like it's also not in scope for the comprehension target, but it works in the iterator:

```py
class C:
    x = 1
    y = [2 for _ in [x]]  # Okay
    z = [x for _ in [1]]  # Error
```

This almost feels like a CPython bug, but maybe it's intentional.

**Comment by MeGaGiGaGon** [comment]
It looks like it's intentional. I asked in the python discord, and ZeroIntensity pointed me to [`PEP709`](https://peps.python.org/pep-0709/). Apparently comprehensions used to operate like functions < 3.12, and that "isolation of scope" was kept even after they got inlined, so it's like expecting this to work
```py
class A:
    x = 1
    def foo():
        print(x)
    foo()
```

**Comment by ntBre** [comment]
Nice, thanks for asking! There's even an example of this in the [docs](https://docs.python.org/3/reference/executionmodel.html#resolution-of-names). And I guess the reason this example works:

```py
class C:
    x = 1
    y = [2 for _ in [x]]  # Okay
```

is given [here](https://docs.python.org/3/reference/expressions.html#displays-for-lists-sets-and-dictionaries):

> However, aside from the iterable expression in the leftmost for clause, the comprehension is executed in a separate implicitly nested scope.

So it's all working as intended, just a bit surprising, at least to me!

**Comment by mikeleppane** [comment]
I will take a look at this.
